### ISSUE TITLE
Truncation is json_encoding the entire payload repeatedly

### ISSUE THREAD

**Issue by passcod** [rollbar/rollbar-php#318]
Found while hunting the bug fixed in #317.

This library, while trying to format the payload for sending, `json_encode`s the entire payload repeatedly. For small inputs, that may seem okay. When the structure is massive, it can eat lots of memory, take lots of time, and eventually crash PHP by trying to allocate more memory than it can.

Here is [Truncation.php](https://github.com/rollbar/rollbar-php/blob/master/src/Truncation/Truncation.php#L26):

```php
foreach (static::$truncationStrategies /* <- 4 elements */ as $strategy) {
   if (!$this->needsTruncating($payload)) { /* <- calls json_encode($payload) */
       break;
   }

   // ...
}
```

"But", you may say, "it's not the same payload, we're modifying it every time!"

Firstly: yes, but you're still `json_encode`-ing a massive object, then a hopefully smaller massive object, then a hopefully even smaller massive object... etc.

And secondly, you're not even doing that! Here's the first strategy:

```php
public function execute(array $payload)
{
    return $payload;
}
```

So, right off the bat here you're encoding the payload, doing nothing with it, and then encoding it again. Let's check out the next strategy:

```php
if (count($frames) <= $range * 2) { /* $range = 75 */
    return $frames;
}
```

So if there's less than 150 frames in the trace, which will be most times, you're doing nothing... and encoding the same massive object again. Next:

```php
foreach (static::getThresholds() /* 3 items */ as $threshold) {
    if (!$this->truncation->needsTruncating($payload)) {
        break;
    }
```

Here you're running the `json_encode` again 1 to 3 times. The first iteration there runs it on the payload that was just passed in, and as per the code in Truncation.php, the encode was _just_ called on it! And if/when you do `break`, the Strategy returns, and the Truncation immediately calls `json_encode` again at the start of the next iteration, on the exact same object.

In the case we were hitting, _every single `json_encode` was called_, because the payload was absolutely humongous. So Rollbar was attempting to encode this beast 7 times, in close succession, and our guess is that GC didn't have time to run and it ate up all the memory. We even set PHP's max memory to **6 GB** and it ate through it all!

(And most frustrating, all this work is for naught as in the end, assuming it skirts under the memory limit and finishes processing, the payload would still be too big, and would just have been silently discarded.)

---

I'm not quite sure how this can be fixed, but at the very least not encoding the same object twice would be nice.

---

If anyone comes looking and finds this, an effective workaround if you know what's causing the bloat is to implement the [`JsonSerializable`](http://php.net/manual/en/jsonserializable.jsonserialize.php) interface and just specify that whatever massive object you found serialises to something tiny (while still providing context). For example, we used this:

```php
public function jsonSerialize()
{
    return get_called_class();
}
```

**Comment by ArturMoczulski** [comment]
@passcod great insights. I will be working on this soon.

**Comment by passcod** [comment]
So, amendment on that, because we found another bug, see #323.

> (And most frustrating, all this work is for naught as in the end, assuming it skirts under the memory limit and finishes processing, the payload would still be too big, and would just have been silently discarded.)

That's false. Upon re-read, we found that if the payload is still too large, then it gets sent anyway. Hooray.

What we also found is that the truncation code assumes the only big parts are going to be the exception message and/or the exception frames. If, say, the massive part is the associated data in e.g. a manual `Rollbar::log` call... the strategies never actually check that, happily try to `json_encode` things over and over again, never actually doing anything because the problem isn't with the frames.

And then we found another spurious `json_encode`. It's in debug code. Look:

```php
$toSend = $this->truncate($toSend);

$this->debugLogger->info(
    "Payload scrubbed and ready to send to ".
    $this->config->getSender()->toString()
);
$this->debugLogger->debug(json_encode($toSend, true));
```

See [that last line](https://github.com/rollbar/rollbar-php/blob/2c50b41/src/RollbarLogger.php#L105)?

Perfectly reasonable... you just want to see the final payload in your debug logs. I'm sure that's very useful. Except... there's no gate on this debug call. Which means that whether or not that debug log is enabled, the payload gets (uselessly, if debug is disabled) `json_encode`d again.
